runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientProduceFactory.java

```diff
@@ -1654,85 +1634,135 @@ private void doEncodeRecordInit(
+                    encodeSlotOffset = PRODUCE_REQUEST_RECORDS_OFFSET_MAX;
+                    encodeSlotLimit = encodeSlotOffset;
+                    encodeableRecordBatchSlotOffset = encodeSlotLimit;
+                }
+
+                short attributes = encodeableRecordBatchTimestampMax == 0L
+                        ? RECORD_BATCH_ATTRIBUTES_NO_TIMESTAMP
+                        : RECORD_BATCH_ATTRIBUTES_NONE;
+                int recordCount = 1;
+                int oldRecordBatchLength = FIELD_OFFSET_RECORD_COUNT - FIELD_OFFSET_LENGTH;
+
+                assert encodeSlot != NO_SLOT;
+                final int maxLimit = encodePool.slotCapacity();
+                final MutableDirectBuffer encodeSlotBuffer = encodePool.buffer(encodeSlot);
+
+                final RecordBatchFW latestRecordBatch = recordBatchRO.tryWrap(
+                    encodeSlotBuffer, encodeableRecordBatchSlotOffset, encodeSlotLimit);
+
+                if (latestRecordBatch != null &&
+                    latestRecordBatch.producerId() == producerId &&
+                    latestRecordBatch.producerEpoch() == producerEpoch &&
+                    latestRecordBatch.baseSequence() <= sequence)
+                {
+                    oldRecordBatchLength = latestRecordBatch.length();
+                    producerId = latestRecordBatch.producerId();
+                    producerEpoch = latestRecordBatch.producerEpoch();
+                    sequence = latestRecordBatch.baseSequence();
+                    recordCount = latestRecordBatch.recordCount() + 1;
+                    attributes = latestRecordBatch.attributes();
+                    encodeableRecordBatchTimestamp = latestRecordBatch.firstTimestamp();
+                    encodeableRecordBatchTimestampMax = latestRecordBatch.maxTimestamp();
+                    encodeSlotLimit = encodeableRecordBatchSlotOffset;
+                    encodeableRecordBatchBytes -= latestRecordBatch.sizeof();
+                }
+
+                sequence = producerId == RECORD_BATCH_PRODUCER_ID_NONE ? RECORD_BATCH_BASE_SEQUENCE_NONE : sequence;
+
+                final int oldEncodeSlotLimit = encodeSlotLimit;
+
+                final RecordBatchFW recordBatch = recordBatchRW.wrap(encodeSlotBuffer, encodeSlotLimit, maxLimit)
+                        .baseOffset(0)
+                        .length(oldRecordBatchLength)
+                        .leaderEpoch(-1)
+                        .magic(RECORD_BATCH_MAGIC)
+                        .crc(0)
+                        .attributes(attributes)
+                        .lastOffsetDelta(recordCount - 1)
+                        .firstTimestamp(encodeableRecordBatchTimestamp)
+                        .maxTimestamp(encodeableRecordBatchTimestampMax)
+                        .producerId(producerId)
+                        .producerEpoch(producerEpoch)
+                        .baseSequence(sequence)
+                        .recordCount(recordCount)
+                        .build();
```

Comment:
Note: the `if` condition above is checking `latestRecordBatch.producerId() == producerId` then assigning `producerId = latestRecordBatch.producerId()` (which is a no-op if they are already same).

Reply:
Suggest we adjust the logic to simplify.

First let's conditionally check if a new record batch is needed (if no current batch, producer differs, etc), and create an empty batch (header) if needed, with record count 0 and record batch length for 0 records.

Then, let's dynamically extend the current record batch to increase the record count by 1, the record batch length by new record bytes, plus the record bytes too. This logic should work whether the record batch was just created (empty, no records yet) or already exists with previously encoded records in the batch.

Modifying the record batch record count and length should be feasible with direct update (like crc) instead of rewriting the record batch (header) entirely.